### DISCUSSION TITLE
DataGrid row selection before command when clicking delete button

### DISCUSSION THREAD

**Discussion by RealityMasque** [Megabit/Blazorise#4942] [General]
Hi,

New to Blazorise, & having a bit of difficulty.  I have a delete button per row, but depending on how fast I click, either the row is selected, the delete icon is selected, or the command is actually invoked.  I've tried a number of potential fixes found here & across the web, but none seem to help.

Help?

Thanks,

Orion

**Comment by David-Moreira** [comment]
Can you please provide us with a reproducable? 
For instance, can you reproduce the issue on our applications? 

https://blazorise.com/docs/extensions/datagrid/features/editing
https://bootstrapdemo.blazorise.com/tests/datagrid/editing/commands

**Comment by RealityMasque** [comment]
your 2nd link seems to work more like what we need.  coming up with code example now - it's proprietary.

**Comment by RealityMasque** [comment]
I can't seem to paste the code here w/o losing some of it

**Comment by RealityMasque** [comment]
&lt;DataGrid TItem="ItemType"
          Data="@SelectedParentItem.ItemList"
@bind-SelectedRow="@SelectedItem"
          DetailRowTrigger="@(Conditions)"
          Responsive
           ShowCaptions="true"
           Editable="@(UserAuthorized)"
           Bordered=true
           ShowPager=true
           PageSize="10" &gt;
     &lt;DataGridColumns&gt;
         &lt;DataGridCommandColumn TItem="ItemType"&gt;
             &lt;EditCommandTemplate&gt;
                 &lt;Tooltip Inline="true" Text="Edit Item"&gt;
                     &lt;Button Color="Color.Warning" Outline=true Clicked="@(e =&gt; EditItem(SelectedItem))"&gt;&lt;Icon Name="IconName.Pen"&gt;&lt;/Icon&gt;&lt;/Button&gt;
                 &lt;/Tooltip&gt;
                 &lt;Tooltip Inline="true" Text="Delete Item"&gt;
                     &lt;Button Color="Color.Danger" Clicked="@(e =&gt; DeleteItem(SelectedItem))" Outline=true&gt;&lt;Icon Name="IconName.Delete"&gt;&lt;/Icon&gt;&lt;/Button&gt;
                 &lt;/Tooltip&gt;
             &lt;/EditCommandTemplate&gt;
             &lt;NewCommandTemplate&gt;
                 &lt;Tooltip Inline="true" Text="New Item"&gt;
                     &lt;Button Color="Color.Secondary" Clicked="@(e =&gt; AddItem())"&gt;&lt;Icon Name="IconName.PlusCircle"&gt;&lt;/Icon&gt;&lt;/Button&gt;
                 &lt;/Tooltip&gt;
             &lt;/NewCommandTemplate&gt;
             &lt;SaveCommandTemplate&gt;
             &lt;/SaveCommandTemplate&gt;
             &lt;DeleteCommandTemplate&gt;
             &lt;/DeleteCommandTemplate&gt;
         &lt;/DataGridCommandColumn&gt;
         &lt;DataGridColumn TItem="ItemType" Field="@nameof(ItemType.Prop1)" Caption="Prop1" /&gt;
         &lt;DataGridColumn TItem="ItemType" Field="@nameof(ItemType.Prop2)" Caption="Prop2"&gt;
             &lt;DisplayTemplate Context="subcontext"&gt;
                 @{
                    var value = (subcontext as ItemType).ObjectType;
                    string output = EnumHelper.GetString(value);
                }
                @(output)
            &lt;/DisplayTemplate&gt;
        &lt;/DataGridColumn&gt;
        &lt;DataGridColumn TItem="ItemType" Field="@nameof(ItemType.Prop3)" Caption="Prop3"&gt;
            &lt;DisplayTemplate Context="subcontext"&gt;
                @{
                    var value = (subcontext as ItemType).Prop3;
                    string output = EnumHelper.GetString(value);
                }
                @(output)
            &lt;/DisplayTemplate&gt;
        &lt;/DataGridColumn&gt;
        &lt;DataGridColumn TItem="ItemType" Field="@nameof(ItemType.Prop4)" Caption="Prop4" /&gt;
        &lt;DataGridColumn TItem="ItemType" Field="@nameof(ItemType.Prop5)" Caption="Prop5" Sortable="true"&gt;
            &lt;DisplayTemplate&gt;
                &lt;Check TValue="bool" Checked="context.Prop5" Disabled=true /&gt;
            &lt;/DisplayTemplate&gt;
        &lt;/DataGridColumn&gt;
    &lt;/DataGridColumns&gt;
    &lt;DetailRowTemplate&gt;
        @{
            var data = (context as ItemType);

            <CardDeck>
                <Card>
                    <CardBody>
                        <strong>Comments1</strong>
                    </CardBody>
                </Card>
                <Card>
                    <CardBody>
                        <strong>Comments<2/strong>
                    </CardBody>
                </Card>
            </CardDeck>
        }
    </DetailRowTemplate>
    <EmptyTemplate>
        <div class="box">
            <i>not added</i>
        </div>
    </EmptyTemplate>
&lt;/DataGrid&gt;

@code {

    [Parameter]
    public ParentItemType SelectedParentItem { get; set; }

    [CascadingParameter]
    private Task>AuthenticationState< authenticationStateTask { get; set; }

    private User _userModel = new();

    public ItemType SelectedItem = new();

    protected override async Task OnInitializedAsync()
    {
        _userModel = UserController.GetUser(email);
    }
	
    private bool showModal = false;

    private void AddItem()
    {
        SelectedItem = new()
            {
                KeyIdentifier = Guid.NewGuid().ToString()
            };
        showModal = true;
    }

    private void EditItem(ItemType anItem)
    {
        SelectedItem = anItem;
        showModal = true;
    }

    private void DeleteItem(ItemType anItem)
    {
        SelectedParentItem.ItemList.Remove(anItem);
    }

    private void SaveKey()
    {
        if (SelectedParentItem.ItemList.Any(x => x.KeyIdentifier == SelectedItem.KeyIdentifier))
        {
            SelectedItem itemToUpdate = SelectedParentItem.ItemList.SingleOrDefault(x => x.KeyIdentifier == SelectedItem.KeyIdentifier);
            SelectedParentItem.ItemList.Remove(itemToUpdate);
        }
        SelectedParentItem.ItemList.Add(SelectedItem);

        HideModal();
    }

    private void HideModal()
    {
        showModal = false;
    }
}

**Comment by RealityMasque** [comment]
@David-Moreira - reproducable provided-ish

**Comment by stsrki** [comment]
hello @RealityMasque can you please try to paste the code again? To properly format it inside of a block code, you need to make it like in markdown specs, see https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by RealityMasque** [reply]
are you unable to read what I put in above, @stsrki ?

**Reply by stsrki** [reply]
I would try to copy/paste it into the app to test it. But there are too many HTML escape codes, and it doesn't work. That why I'm saying it is best to give us a reproducible code. It saves time for all of us, and we can help you sooner.

**Reply by RealityMasque** [reply]
when I copy & paste in a code block, the post doesn't render everything

**Reply by RealityMasque** [reply]
@stsrki  ^

**Reply by stsrki** [reply]
That's because you need to make sure the code is placed inside of "code blocks". In markdown, the code block is writer with three backticks

eg.

![image](https://github.com/Megabit/Blazorise/assets/900302/b1bddb9a-c77a-42bd-a9b5-a89e41dec982)

**Comment by David-Moreira** [comment]
Hello,

this should fix it:
```
<EditCommandTemplate>
                 <Div @onclick:stopPropagation>
                     <Tooltip Inline="true" Text="Edit Item">
                         <Button Color="Color.Warning" Outline=true Clicked="@(e => EditItem(SelectedItem))"><Icon Name="IconName.Pen"></Icon></Button>
                     </Tooltip>
                     <Tooltip Inline="true" Text="Delete Item">

                         <Button Color="Color.Danger" Clicked="@(e => DeleteItem(SelectedItem))" Outline=true><Icon Name="IconName.Delete"></Icon></Button>
                     </Tooltip>
                 </Div>
             </EditCommandTemplate>
```
Making it so the click is not propagated up to the row. let us know if this helped.

**Reply by David-Moreira** [reply]
@RealityMasque Any chance you've tried this yet? May you let us know if this helped you?

**Reply by RealityMasque** [reply]
sorry - had a bunch of other stuff to get to.  is there a C# version of this property/attribute?

**Reply by David-Moreira** [reply]
What do you mean? That's c# / razor syntax : `@onclick:stopPropagation`

**Reply by RealityMasque** [reply]
Doesn't this do the same thing?  PreventDefaultOnSubmit, or is that only for submit buttons?

I tried @onclick:stopPropagation, but that just kept the delete button clicked event from being called after the 1st rows was deleted.

**Reply by David-Moreira** [reply]
Ah well... I didn't fix your code logic wise. You are deleting the current SelectedItem not the row item.

**Reply by David-Moreira** [reply]
It should work as you'd like after fixing that.

**Reply by stsrki** [reply]
The PreventDefaultOnSubmit is meant to prevent the form submit when the `Button` is placed inside of a `Form` component.

**Comment by RealityMasque** [comment]
> Ah well... I didn't fix your code logic wise. You are deleting the current SelectedItem not the row item.

without @onclick:stopPropagation, when I click the delete buttons infrequently, every item gets deleted.  with @onclick:stopPropagation, some of the items remain, & the delete button remains after being clicked & stop firing the DeleteKey() method.

**Reply by David-Moreira** [reply]
Did you fix your logic to not delete the **SelectedItem**? But the row item instead? If the row doesn't get clicked the `SelectedItem` does not get updated with that row's Item. You need to use the `context.Item` to get & delete the row item.

`Clicked="@(e => DeleteItem(SelectedItem))"`

**Comment by RealityMasque** [comment]
we decided to go with using an "are you sure?" modal to avoid the clicking delete repeatedly
